src/equations/serre_green_naghdi_1d.jl

```diff
@@ -97,62 +181,204 @@ function create_cache(mesh,
+                 M_h_p_h_bx2, M_h3_3, M_h2_bx,
+                 D1, D1mat_minus, factorization)
+    else
+        if D1 isa FourierDerivativeOperator
+            D1mat = Matrix(D1)
+
+            cache = (; h, h_x, v_x, h_hpb_x, b, b_x, hv_x, v2_x,
+                     h2_v_vx_x, h_vx_x, p_h, p_x, tmp,
+                     M_h_p_h_bx2, M_h3_3, M_h2_bx,
+                     D1, D1mat)
+        else
+            D1mat = sparse(D1)
+
+            system_matrix = let cache = (; M_h_p_h_bx2, M_h3_3, M_h2_bx)
+                assemble_system_matrix!(cache, h, b_x,
+                                        D1, D1mat, equations)
+            end
+            factorization = cholesky(system_matrix)
+
+            cache = (; h, h_x, v_x, h_hpb_x, b, b_x, hv_x, v2_x,
+                     h2_v_vx_x, h_vx_x, p_h, p_x, tmp,
+                     M_h_p_h_bx2, M_h3_3, M_h2_bx,
+                     D1, D1mat, factorization)
+        end
+    end
+
+    if equations.bathymetry_type isa BathymetryVariable
+        psi = zero(h)
+        cache = (; cache..., psi)
+    end
+
+    return cache
+end
+
+function assemble_system_matrix!(cache, h, D1, D1mat,
+                                 equations::SerreGreenNaghdiEquations1D{BathymetryFlat})
+    (; M_h, M_h3_3) = cache
+
+    @. M_h = h
+    scale_by_mass_matrix!(M_h, D1)
+    @. M_h3_3 = (1 / 3) * h^3
+    scale_by_mass_matrix!(M_h3_3, D1)
+
+    # Floating point errors accumulate a bit and the system matrix
+    # is not necessarily perfectly symmetric but only up to
+    # round-off errors. We wrap it here to avoid issues with the
+    # factorization.
+    return Symmetric(Diagonal(M_h) + D1mat' * Diagonal(M_h3_3) * D1mat)
+end
+
+# variable bathymetry
+function assemble_system_matrix!(cache, h, b_x, D1, D1mat,
+                                 equations::SerreGreenNaghdiEquations1D)
+    (; M_h_p_h_bx2, M_h3_3, M_h2_bx) = cache
+
+    if equations.bathymetry_type isa BathymetryMildSlope
+        factor = 0.75
+    elseif equations.bathymetry_type isa BathymetryVariable
+        factor = 1.0
+    end
+    @. M_h_p_h_bx2 = h + factor * h * b_x^2
+    scale_by_mass_matrix!(M_h_p_h_bx2, D1)
+    inv3 = 1 / 3
+    @. M_h3_3 = inv3 * h^3
+    scale_by_mass_matrix!(M_h3_3, D1)
+    @. M_h2_bx = 0.5 * h^2 * b_x
+    scale_by_mass_matrix!(M_h2_bx, D1)
+    # Floating point errors accumulate a bit and the system matrix
+    # is not necessarily perfectly symmetric but only up to
+    # round-off errors. We wrap it here to avoid issues with the
+    # factorization.
+    return Symmetric(Diagonal(M_h_p_h_bx2)
+                     +
+                     D1mat' * (Diagonal(M_h3_3) * D1mat
+                               -
+                               Diagonal(M_h2_bx))
+                     -
+                     Diagonal(M_h2_bx) * D1mat)
+end
+
+function solve_system_matrix!(dv, system_matrix, rhs,
+                              ::SerreGreenNaghdiEquations1D,
+                              D1, cache)
```

Comment:
This looks quite nice. For the future we could even think of reusing this function also for the other equations. For the Svärd-Kalisch equations I think we could also use it already, but not for BBM-BBM equations because we use LU instead of Cholesky decomposition, which will hopefully change some day, see #122.